distributed/actor.py

```diff
@@ -227,34 +253,61 @@ class ActorFuture:
     Actor
     """
 
-    def __init__(self, q, io_loop, result=None):
-        self.q = q
-        self.io_loop = io_loop
-        if result:
-            self._cached_result = result
-        self.status = "pending"
+    @abc.abstractmethod
+    def __await__(self):
+        pass
+
+    @abc.abstractmethod
+    def result(self, timeout=None):
+        pass
+
+    @abc.abstractmethod
+    def done(self):
+        pass
+
+    def __repr__(self):
+        return "<ActorFuture>"
+
+
+class _EagerActorFuture(ActorFuture):
+    """Future to an actor's method call when an actor calls another actor on the same worker"""
+
+    def __init__(self, result):
+        self._result = result
+
+    def __await__(self):
+        return self._result
+        yield
+
+    def result(self, timeout=None):
+        return self._result
+
+    def done(self):
+        return True
+
+
+class _ActorFuture(ActorFuture):
```

Comment:
These are marked as private classes, but they *do* make their way into the user's hand through normal usage of the API (despite the lying repr). I'd rather they be public, and perhaps name the abstract base something like `ActorFutureBase`.

Reply:
It looks like the abstract base class should be private instead of this one. As long as we underscore all attributes it sounds like a good idea to make these public

Reply:
I don't want to make the constructors of _ActorFuture or _EagerActorFuture public, and particularly I don't want to make:
```
if isinstance(fut, actor._EagerActorFuture):
    behave(fut)
```
public

Reply:
FWIW, in dask/dask, the problem of a "non-public constructor" is handled socially, via docstring: https://github.com/dask/dask/blob/193998cef5f3f2c7e22fd5d2eb2f7b0b41b35d63/dask/dataframe/core.py#L4055-L4062 .

Reply:
Agreed with Florian and Ian here. Non-underscored names and a message in the docstring saying "users should not construct this class directly" seems best to me. I'd also prefer the `ActorFuture` ABC to be public as well—it could be helpful to library developers or users as a type annotation or `isinstance` check, plus might be a better thing to document (if we add docstrings) than the various concrete subclasses.